Sources/NIOPosix/Bootstrap.swift

```diff
@@ -2281,11 +2281,18 @@ extension NIOPipeBootstrap {
 
             inputFileHandle = input.flatMap { NIOFileHandle(descriptor: $0) }
             outputFileHandle = output.flatMap { NIOFileHandle(descriptor: $0) }
-            channel = try PipeChannel(
-                eventLoop: eventLoop as! SelectableEventLoop,
-                inputPipe: inputFileHandle,
-                outputPipe: outputFileHandle
-            )
+            do {
+                channel = try PipeChannel(
+                    eventLoop: eventLoop as! SelectableEventLoop,
+                    inputPipe: inputFileHandle,
+                    outputPipe: outputFileHandle
+                )
+            } catch {
+                // Release file handles back to the caller in case of failure.
+                _ = try? inputFileHandle?.takeDescriptorOwnership()
```

Comment:
This is doing the opposite of the comment. It takes the ownership instead of leaving it with the caller.

If this method throws it means that NIO has _not_ taken ownership, so the user keeps it.

This change doesn't seem right, the old code seems correct. What's the issue with it?

Reply:
We take ownership of the handle when we wrap it in `NIOFileHandle` on lines 2282 and 2283. If the `PipeChannel` constructor throws, it will not close the `NIOFileHandle`s (as it isn't fully constructed at the time). That will cause them to hit their `deinit` (as they go out of scope here) and crash.

Put another we: we _took_ ownership at 2282: if we throw here, we have to "un-take" it, and this is how you spell that on `NIOFileHandle`.